File: path_sum.py
# Time Complexity :O(N)
# Space Complexity :O(LogN)
# Did this code successfully run on Leetcode : Yes

# Any problem you faced while coding this :

# Your code here along with comments explaining your approach


class Solution:
    # result=[]
    # def list_append(self, val):
    #     l=[]
    #     l.append(val)
    #     return l

    def helper(self, root, tar,c_sum, path, result):
        #base
        if(root==None):
            return
        if(root.left==None and root.right==None):
            c_sum+=root.val
            print(c_sum, root.val)
            if(c_sum==tar):
                path.append(root.val)
                result.append(path)
            return result
        #logic
        c_sum+=root.val        
        self.helper(root.left, tar, c_sum, path+[root.val],result)
        self.helper(root.right, tar, c_sum,path+[root.val], result)
        if(len(path)):
            path.pop()
        return result


    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        # self.result=[]
        return self.helper(root, targetSum,0, [],[])
        print(self.result)
        return self.result

File: PathSum.cs
// Time Complexity : O(n^2)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private IList<IList<int>> result;
    public IList<IList<int>> PathSum(TreeNode root, int targetSum) {
        result = new List<IList<int>>();
        var li = new List<int>();        
        helper(root, 0, li, targetSum);
        return result;        
    }

    public void helper(TreeNode root, int currSum, List<int> li, int targetSum)
    {
        if(root == null) return;

        li.Add(root.val);
        currSum += root.val;
                
        helper(root.left, currSum, new List<int>(li), targetSum);

        if(root.left == null && root.right == null && currSum == targetSum)
        {            
            result.Add(li);                    
        }
        helper(root.right, currSum, new List<int>(li), targetSum);
                     
    }
}
